common/deliverclient/block_verification.go

```diff
@@ -0,0 +1,350 @@
+		verifierAssembler:   bva,
+		sigVerifierFunc:     verifierFunc,
+		lastBlockHeader:     &common.BlockHeader{Number: lastBlockNumber},
+		lastBlockHeaderHash: lastBlockHeaderHash,
+		logger:              lg,
+	}
+
+	return a, nil
+}
+
+func (a *BlockVerificationAssistant) Clone() CloneableUpdatableBlockVerifier {
+	c := &BlockVerificationAssistant{
+		channelID:           a.channelID,
+		verifierAssembler:   a.verifierAssembler,
+		sigVerifierFunc:     a.sigVerifierFunc,
+		configBlockHeader:   a.configBlockHeader,
+		lastBlockHeader:     a.lastBlockHeader,
+		lastBlockHeaderHash: a.lastBlockHeaderHash,
+		logger:              a.logger,
+	}
+	return c
+}
+
+// UpdateConfig sets the config by which blocks are verified. It is assumed that this config block had already been
+// verified using the VerifyBlock method immediately prior to calling this method.
+func (a *BlockVerificationAssistant) UpdateConfig(configBlock *common.Block) error {
+	configTx, err := protoutil.ExtractEnvelope(configBlock, 0)
+	if err != nil {
+		return errors.WithMessage(err, "error extracting envelope")
+	}
+
+	payload, err := protoutil.UnmarshalPayload(configTx.Payload)
+	if err != nil {
+		return errors.WithMessage(err, "error umarshaling envelope to payload")
+	}
+
+	if payload.Header == nil {
+		return errors.New("missing channel header")
+	}
+
+	chdr, err := protoutil.UnmarshalChannelHeader(payload.Header.ChannelHeader)
+	if err != nil {
+		return errors.WithMessage(err, "error unmarshalling channel header")
+	}
+
+	if chdr.GetChannelId() != a.channelID {
+		return errors.New("config block channel ID '%s' does not match expected: '%s'")
+	}
+
+	configEnvelope, err := configtx.UnmarshalConfigEnvelope(payload.Data)
+	if err != nil {
+		return errors.WithMessage(err, "error umarshaling config envelope from payload data")
+	}
+
+	verifierFunc, err := a.verifierAssembler.VerifierFromConfig(configEnvelope, chdr.GetChannelId())
+	if err != nil {
+		return errors.WithMessage(err, "error creating verifier function")
+	}
+
+	a.configBlockHeader = configBlock.Header
+	a.lastBlockHeader = configBlock.Header
+	a.lastBlockHeaderHash = protoutil.BlockHeaderHash(configBlock.Header)
+	a.sigVerifierFunc = verifierFunc
+
+	return nil
+}
+
+// VerifyBlock checks block integrity and its relation to the chain, and verifies the signatures.
+func (a *BlockVerificationAssistant) VerifyBlock(block *common.Block) error {
+	if err := a.verifyHeader(block); err != nil {
+		return err
+	}
+
+	// Verify channel ID
+	channelID, err := protoutil.GetChannelIDFromBlock(block)
+	if err != nil {
+		return errors.Wrapf(err, "failed getting channel id from block number [%d] on channel [%s]", block.Header.Number, a.channelID)
+	}
+
+	if channelID != a.channelID {
+		return errors.Errorf("invalid block's channel id. Expected [%s]. Given [%s]", a.channelID, channelID)
+	}
+
+	if err = a.verifyMetadata(block); err != nil {
+		return err
+	}
+
+	dataHash, err := protoutil.BlockDataHash(block.Data)
```

Comment:
this should be first, before we check the signature.

Reply:
The signature is verified after this